plugins/plugin-jenkins/codenvy-plugin-jenkins-webhooks/src/main/java/com/codenvy/plugin/jenkins/webhooks/JenkinsWebhookManager.java

```diff
@@ -0,0 +1,140 @@
+     * If the factory for given commit is already exist from previous requests, this step will be skipped.
+     * 2. Update Jenkins job description with build failed factory url.
+     *
+     * @param jenkinsEvent
+     *         {@link JenkinsEvent} object that contains information about failed Jenkins build
+     * @throws ServerException
+     *         when error occurs during handling failed job event
+     */
+    void handleFailedJobEvent(JenkinsEvent jenkinsEvent) throws ServerException {
+        JenkinsConnector jenkinsConnector = jenkinsConnectorFactory.create(jenkinsEvent.getJenkinsUrl(), jenkinsEvent.getJobName())
+                                                                   .updateUrlWithCredentials();
+        try {
+            String commitId = jenkinsConnector.getCommitId(jenkinsEvent.getBuildId());
+            String repositoryUrl = jenkinsEvent.getRepositoryUrl();
+            Optional<Factory> existingFailedFactory = getExistingFailedFactoryIfPresent(repositoryUrl, commitId);
+            Factory failedFactory = existingFailedFactory.isPresent() ? existingFailedFactory.get() :
+                                    createFailedFactory(jenkinsConnector.getFactoryId(),
+                                                        repositoryUrl,
+                                                        commitId);
+            jenkinsConnector.addFailedBuildFactoryLink(baseUrl.substring(0, baseUrl.indexOf("/api")) + "/f?id=" + failedFactory.getId());
+        } catch (IOException | NotFoundException | ConflictException e) {
+            LOG.error(e.getLocalizedMessage());
+            throw new ServerException(e.getLocalizedMessage());
+        }
+    }
+
+    private Optional<Factory> getExistingFailedFactoryIfPresent(String repositoryUrl, String commitId) throws NotFoundException,
+                                                                                                              ServerException {
+        return getUserFactories().stream()
+                                 .filter(f -> f.getWorkspace()
+                                               .getProjects()
+                                               .stream()
+                                               .anyMatch(project -> repositoryUrl.equals(project.getSource().getLocation()) &&
+                                                                    commitId.equals(project.getSource()
+                                                                                           .getParameters()
+                                                                                           .get("commitId"))))
+                                 .findFirst();
+    }
+
+    private List<Factory> getUserFactories() throws NotFoundException, ServerException {
+        List<Factory> factories = new ArrayList<>();
+        List<Factory> request;
```

Comment:
Actually it is response but not request :smile: I'd name it as `factoriesPage` or something like that.